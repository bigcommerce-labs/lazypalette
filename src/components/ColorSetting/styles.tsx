import styled from 'styled-components';

export const Container = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

interface SelectedColorProps {
    color: string;
}
export const SelectedColor = styled.div.attrs<SelectedColorProps>({}) `
  background: ${(props: SelectedColorProps) => `${props.color}`};
  border: 1px solid ${({ theme }) => theme.colors.stroke};
  border-radius: 3px;
  box-sizing: border-box;
  cursor: pointer;
  flex-shrink: 0;
  height: 1.5rem;
  width: 1.5rem;
`;

export const ColorText = styled.div`
    color: ${({ theme }) => theme.colors.secondaryText};
    flex-grow: 1;
    font-size: ${({ theme }) => theme.typography.fontSize.smaller};
    margin-right: 0.5rem;
    text-align: right;
`;

export const Label = styled.label`
  color: ${({ theme }) => theme.colors.primaryText};
`;
