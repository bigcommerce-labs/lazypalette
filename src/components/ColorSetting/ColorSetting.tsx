import React, { PureComponent } from 'react';
import { ColorResult } from 'react-color';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import uuid from 'uuid';

import { ThemeConfigChange } from '../../actions/theme';
import { openColorPicker, ColorPickerContent, OpenColorPickerArgs, OpenUIWindowAction } from '../../actions/uiWindow';

import { ColorText, Container, Label, SelectedColor } from './styles';

interface ColorSettingProps {
    color?: string;
    inputId?: string;
    label?: string;
    name: string;
    testId?: string;
    onChange?(configChange: ThemeConfigChange): void;
    openColorPicker(content: OpenColorPickerArgs): OpenUIWindowAction<ColorPickerContent>;
}

interface ColorSettingState {
    color: string;
    focus: boolean;
    inputId: string;
    previousColor?: string;
    testId?: string;
}

const defaultColor = '#000000';

export class ColorSetting extends PureComponent<ColorSettingProps, ColorSettingState> {
    readonly state: ColorSettingState = {
        color: this.props.color || defaultColor,
        focus: false,
        inputId: this.props.inputId || uuid(),
        previousColor: this.props.color || defaultColor,
        testId: this.props.testId,
    };

    handleChange = (color: ColorResult) => {
        const { name, onChange } = this.props;

        this.setState({ color: color.hex });

        if (onChange) {
            onChange({ setting: { id: name, type: 'color' }, value: color.hex });
        }
    };

    handleOnClose = () => this.setState({ focus: false });

    handleClick = () => {
        this.setState(prevState => ({ previousColor: prevState.color }));

        const { left, width, top } =
            (ReactDOM.findDOMNode(this.refs[this.state.inputId]) as Element).getBoundingClientRect();

        const position = { x: left + (width * 3), y: top };
        this.setState({ focus: true });
        this.props.openColorPicker({
            color: this.state.color,
            onChange: this.handleChange,
            onClose: this.handleOnClose,
            position,
        });
    };

    componentDidUpdate(prevProps: ColorSettingProps) {
        const { color } = this.props;

        if (color && color !== prevProps.color) {
            this.setState({ color });
        }
    }

    render() {
        return (
            <Container>
                <Label>{this.props.label}
                </Label>
                <ColorText>{this.state.color.toUpperCase()}</ColorText>
                <SelectedColor
                    focus={this.state.focus}
                    color={this.state.color}
                    ref={this.state.inputId}
                    onClick={this.handleClick}
                    {...(this.state.testId ? {'data-test-id': this.state.testId} : {})}
                />
            </Container>
        );
    }
}

const mapDispatchToProps = {
    openColorPicker,
};

export default connect(null, mapDispatchToProps)(ColorSetting);
