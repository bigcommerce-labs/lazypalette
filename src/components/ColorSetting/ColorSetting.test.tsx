import { shallow } from 'enzyme';
import React from 'react';
import createMockStore from 'redux-mock-store';

import ColorSetting from './ColorSetting';

const mockStore = createMockStore([]);

it('renders', () => {
    const store = mockStore({});

    const colorSetting = shallow(
        <ColorSetting inputId="blah" name="blah2" />,
        { context: { store } }).dive();
    expect(colorSetting).toMatchSnapshot();
});

it('renders an initial color if provided', () => {
    const store = mockStore({});
    const colorSetting = shallow(
        <ColorSetting inputId="blah" name="blah2" color="#fff" />,
        { context: { store } }).dive();
    expect(colorSetting).toMatchSnapshot();
});
