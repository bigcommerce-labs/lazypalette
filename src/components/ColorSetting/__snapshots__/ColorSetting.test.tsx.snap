// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<styled.div>
  <styled.label />
  <styled.div>
    #000000
  </styled.div>
  <styled.div
    color="#000000"
    focus={false}
    onClick={[Function]}
  />
</styled.div>
`;

exports[`renders an initial color if provided 1`] = `
<styled.div>
  <styled.label />
  <styled.div>
    #FFF
  </styled.div>
  <styled.div
    color="#fff"
    focus={false}
    onClick={[Function]}
  />
</styled.div>
`;
